File: my-folder/0040-combination-sum-ii/solution.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        def dfs(i, find, combo):
            if find == target:
                res.append(combo.copy())
                return 
            
            if find > target or i == len(candidates):
                return 

            combo.append(candidates[i])
            dfs(i + 1, find + candidates[i], combo)
            combo.pop()
            while i + 1 < len(candidates) and candidates[i] == candidates[i + 1]:
                i += 1
            dfs(i + 1, find, combo)

        dfs(0, 0, [])
        return res


        

File: my-folder/0040-combination-sum-ii/solution.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(candidates);
        updateCombinationSums(0, candidates, new ArrayList<>(), 0, target, result);
        return result;
    }

    private void updateCombinationSums(int idx, int[] candidates, List<Integer> curr, int currSum, int target, List<List<Integer>> result){

        if(currSum == target){
            result.add(new ArrayList<>(curr));
            return;
        }

        if(currSum > target) return;
        
        for(int i = idx; i < candidates.length; i++){
            // skipping duplicates
            if(i > idx && candidates[i] == candidates[i - 1]) continue;

            curr.add(candidates[i]);
            updateCombinationSums(i + 1, candidates, curr, currSum + candidates[i], target,     result);
            curr.remove(curr.size() - 1);

        }
    }
}
